adding c++ txn sample

andreimatei opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/146)
<!-- Reviewable:end -->

RaduBerinde commented on _includes/app/txn-sample.cpp:
```diff
@@ -0,0 +1,74 @@
+    pqxx::connection c("postgresql://root@localhost:26257");
+    string create_stmts(R"(
+		  CREATE DATABASE bank;
+      SET DATABASE = bank;
```
alignment. I would just separate each statement on a `stmts += ""` line

But it would be nice to encourage not putting things in a single txn unless they really need to be.. We could exec()` and `commit()` each one here.

andreimatei replied:
this is a "nontransaction", see below :)

RaduBerinde replied:
ah, lol. then why does it need a `commit()` call?

andreimatei replied:
it just does man, this library is very finicky about the state of all its objects

RaduBerinde replied:
are you sure it doesn't work without it? I did some quick googling and the docs I found say "Any query executed in a nontransaction is committed immediately, and neither Commit() nor Abort() has any effect." and the example I found has no commit call.

RaduBerinde replied:
if we do need it, maybe add a comment saying it doesn't actually do anything

andreimatei replied:
Without the commit, you get `Started transaction<READ COMMITTED> while nontransaction still active` afterwards. FWIW I've ran into other simply broken examples of using this lib, even on their website.
Added a note and fixed the alignment.

knz commented on _includes/app/txn-sample.cpp:
```diff
@@ -0,0 +1,74 @@
+    pqxx::dbtransaction *tx, int from, int to, float amount) {
+  // Read the balance.
+  pqxx::result r = tx->exec(
+      "SELECT balance FROM accounts WHERE id = " + to_string(from));
```
I'd recommend checking there's at least 1 row in there.

andreimatei replied:
added an assertion

knz commented on _includes/app/txn-sample.cpp:
```diff
@@ -0,0 +1,74 @@
+int main() {
+  try {
+    pqxx::connection c("postgresql://root@localhost:26257");
+    string create_stmts(R"(
```
What is this R prefix to the string? It doesn't seem standard.

andreimatei replied:
raw strings bro :sunglasses: 
http://en.cppreference.com/w/cpp/language/string_literal

knz commented on _includes/app/txn-sample.cpp:
```diff
@@ -0,0 +1,74 @@
+      // Swallow "transaction restart" errors; the transaction will be retried.
+      // Unfortunately libpqxx doesn't give us access to the error code, so we
+      // do string matching to identify retriable errors.
+      if (string(e.base().what()).find("restart transaction:") == string::npos) {
```
why do you need `.base()` here?

andreimatei replied:
you need to do this to get to the std::exception. That's how you get the error message, they say it in the class docs.

andreimatei commented:
@knz the thinking on the back-off delay has evolved a bit - the plan is to do all the waiting in the `store` based on some signaling mechanism - you'll be woken when the intent that's blocking you has gone away and so you can either proceed (if the other guy rolled back) or immediately restart with a higher timestamp.

jseldess commented:
@knz, @RaduBerinde, everything look good to go now?

knz commented:
:lgtm:

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[_includes/app/txn-sample.cpp, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/146#-KDfiiv0J-wHkezJ5ijc:-KDfiiv0J-wHkezJ5ijd:-230574784) ([raw file](https://github.com/cockroachdb/docs/blob/7928bdfae427015d1273015c859cb898ce8ed7d8/_includes/app/txn-sample.cpp#L63)):*
stutter. also add a colon after Note

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/docs/146#-:-KDfipnCP8anI3YkSeP3:1417088721)*
<!-- Sent from Reviewable.io -->

jseldess commented on _includes/app/txn-sample.cpp:
```diff
@@ -0,0 +1,76 @@
+    string create_stmts(R"(
+      CREATE DATABASE bank;
+      SET DATABASE = bank;
+      CREATE TABLE IF NOT EXISTS bank.accounts (id INT PRIMARY KEY, balance DECIMAL);
```
Since you `SET DATABASE = bank`, why do you need `bank.accounts` in `CREATE TABLE` and `INSERT`? Can you just use `accounts`?

RaduBerinde commented:
@jseldess I thought the plan was to switch to use INT instead of DECIMAL and float?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/docs/146#-:-KDfz6UX3V0xnP545Zk8:2022372107)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@RaduBerinde, yes, that's the plan. Andrei just hasn't made those updates yet.

spencerkimball commented on _includes/app/txn-sample.cpp:
```diff
@@ -0,0 +1,76 @@
+    } catch (const pqxx::pqxx_exception& e) {
+      // Swallow "transaction restart" errors; the transaction will be retried.
+      // Unfortunately libpqxx doesn't give us access to the error code, so we
+      // do string matching to identify retriable errors.
```
s/retriable/retryable/ ...?  Neither is correct I guess, but `retryable` wins the Google battle 2 to 1.

andreimatei commented:
Review status: 1 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[_includes/app/basic-sample.cpp, line 19 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/146#-KDxzJhC2Vk_RS0308oQ:-KDxzJhC2Vk_RS0308oR:1280506825) ([raw file](https://github.com/cockroachdb/docs/blob/65383ddbb73175ee0dd81311979299952af78f41/_includes/app/basic-sample.cpp#L19)):*
fixed spaces throughout this file

---

*[_includes/app/txn-sample.cpp, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/146#-KDe4ily4YIx8C9El8un-r2-44:-KDxwjACDK2TM0iQIgT0:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/7928bdfae427015d1273015c859cb898ce8ed7d8/_includes/app/txn-sample.cpp#L44)):*
Done.

---

*[_includes/app/txn-sample.cpp, line 59 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/146#-KDe4ily4YIx8C9El8un-r2-59:-KDxxo0wnxRdY88ncvvv:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/7928bdfae427015d1273015c859cb898ce8ed7d8/_includes/app/txn-sample.cpp#L59)):*
Done.

---

*[_includes/app/txn-sample.cpp, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/docs/146#-KDfiiv0J-wHkezJ5ijc:-KDxwuz3y5c_GXQU80dY:-499146009) ([raw file](https://github.com/cockroachdb/docs/blob/7928bdfae427015d1273015c859cb898ce8ed7d8/_includes/app/txn-sample.cpp#L63)):*
Done.

---

*[_includes/app/txn-sample.cpp, line 15 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/docs/146#-KDxvzoKMg6cwiikhhp0:-KDxvzoLZWdVke4w3IS9:-617818747) ([raw file](https://github.com/cockroachdb/docs/blob/65383ddbb73175ee0dd81311979299952af78f41/_includes/app/txn-sample.cpp#L15)):*
i've reverted this from `auto` to the actual type. The default (according to the Google style guide, for example) is to spell out the type unless you can't, or it's too long, or if you're very lazy.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/docs/146)*
<!-- Sent from Reviewable.io -->
